[rahuljai-05] iP

rahuljai-05 opened the pull request.

rahuljai-05 committed:
Level-1

Add functionality to greet and bid farewell to user

rahuljai-05 committed:
Enable chatbot to echo user commands

Modified the chatbot to repeat user input
and exit when the user says "bye".
This change allows the chatbot to mimic user commands
as a basic interactive feature.

rahuljai-05 committed:
Add task management functionality to Fenix

Added the ability for Fenix to store and list user tasks.
Improved structure by separating concerns
with methods for greeting and bidding farewell,
accepting, getting and processing user input,
storing tasks and showing all tasks.
Modified the output format to differentiate between
user commands and Fenix's responses when modifications
are made.

rahuljai-05 committed:
Implement task completion and marking tasks as undone in Fenix

Introduced the ability to mark tasks as done or not done.
Added a new Task class to encapsulate task-related data and behavior.
Enhanced the Fenix class to handle task marking and unmarking commands.
Updated the task list display to reflect task completion status.
Implemented input validation for task indexing to ensure valid commands.

sarahchow03 commented on src/main/java/Fenix.java:
```diff
@@ -0,0 +1,124 @@
+        case "list":
+            showAllTasks(false);
+            break;
+        case "mark":
+            if (words.length > 1 && isValidIndex(words[1])) {
+                markTaskAsDone(Integer.parseInt(words[1]) - 1);
+            } else {
+                System.out.println(
+                        "Please provide a valid task number to mark");
+            }
+            break;
+        case "unmark":
+            if (words.length > 1 && isValidIndex(words[1])) {
+                unmarkTaskAsDone(Integer.parseInt(words[1]) - 1);
+            } else {
+                System.out.println(
+                        "Please provide a valid task number to unmark");
+            }
```
Cases "mark" and "unmark" can be abstracted into mark and unmark functions.

sarahchow03 commented on src/main/java/Fenix.java:
```diff
@@ -0,0 +1,124 @@
+
+public class Fenix {
+
+    public static final int HORIZONTAL_LINE_USER_COMMAND_LENGTH = 60;
+    public static final int HORIZONTAL_LINE_FENIX_MODIFICATION_LENGTH = 52;
```
Good use of constants to prevent the use of magic numbers in the code base. Excellent job!

sarahchow03 commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,30 @@
+        return isDone;
+    }
+
+    public void markAsDone() {
+        this.isDone = true;
+    }
```
Output of markTaskAsDone() in main can be placed in this function. Similarly, this can be done for unmarkAsDone().

sarahchow03 reviewed (commented):
Overall, function names and variables have been appropriately named. You could work on the documentation of code through the use of comments, so that it is more readable.

Good job!

sarahchow03 commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,30 @@
+        this.isDone = false;
+    }
+
+    // Override
```
nit: Annotation of Override is incorrect and should be @Override instead.

iaso1774 commented on src/main/java/Fenix.java:
```diff
@@ -0,0 +1,124 @@
+        switch (command) {
+        case "bye":
+            bidFarewell();
+            return;
```
If there isn't a break statement, there is supposed to be a //Fallthrough comment, although I'm not sure if this applies if you instead have a return statement.

iaso1774 commented on src/main/java/Fenix.java:
```diff
@@ -0,0 +1,124 @@
+
+    public void showAllTasks(boolean isModified) {
+        String modifiedString = (isModified ? "\t" : "");
+        for (int i = 0; i < taskArray.length && taskArray[i] != null; i += 1) {
```
Use i++ instead of i += 1

iaso1774 commented on src/main/java/Fenix.java:
```diff
@@ -0,0 +1,124 @@
+import java.util.Scanner;
+
+public class Fenix {
+
```
There shouldn't be a blank line here.

iaso1774 reviewed (commented):
I could only find a few minor coding standard errors. Nice job!!!

DucPhong135 commented on src/main/java/Fenix.java:
```diff
@@ -0,0 +1,17 @@
+        String request = "What can I do for you ?";
+        String farewell = "Bye. Hope to see you again soon!";
+        System.out.println(horizontalLine);
+        System.out.println(greeting + "\n" + request);
```
Should avoid writing everything on a main method, divide into different method for easy manage.
There are no spaces so it is hard to keep up with the workflow.
Try to not using the \n in your print statement, instead use another print statement to show that those two are on different lines.

DucPhong135 reviewed (commented):
Overall, since you have't finish yet, I cannot give further comments on your work

rahuljai-05 committed:
Add support for ToDo, Deadline, and Event task types and refactor using inheritance to improve code quality

- Introduced three new task types: ToDo, Deadline, and Event.
  - ToDo: Tracks tasks without a specific date or time.
  - Deadline: Tracks tasks that need to be completed by a certain date.
  - Event: Tracks tasks that span a start and end date or time.
  - Enhanced the `list` command to display tasks with the appropriate type labels and deadlines.
- Refactored task structure using inheritance:
  - Created a base `Task` class to encapsulate common behavior for all task types (e.g., `markAsDone`, `unmarkAsDone`, and status tracking).
  - `ToDo`, `Deadline`, and `Event` classes inherit from `Task`, reusing common functionality and reducing code duplication.
  - Enhanced the `Deadline` and `Event` classes to handle date and time-specific functionality, while leveraging shared behavior from `Task`.
- Improved code quality:
  - Modularized input processing logic into helper methods (new `markAsDone`, `unmarkAsDone`, and `processTasks` methods).
  - Enhanced error handling to provide clear feedback for invalid user input.

rahuljai-05 committed:
Handle errors

Added support to handle the following errors:
- No task provided for addition (throws error in case of deadline and event types as
  you cannot split a null task string to get the deadlines - split error)
- Task input format is erroneous (not enough /'s in the case of deadline and event task types -
  array index out of bounds error when you try to access the 'deadlines')

Added support to respond with corresponding warning messages
for the following invalid input cases:
- Invalid command
- Empty task name field (in the case of deadline and event task types)

rahuljai-05 committed:
Clean up code and add new Fenix Exception class

Code cleanup:
- Organised all the constant strings into a new interface
  which is implemented by the Fenix class
- Decomposed some of the longer return strings into
  a combination of smaller, more descriptive chunks
  using variables

Excetptions added:
- Added a new Fenix Exception class to handle invalid
  inputs which do not necessarily cause errors
- Added extra try-catch blocks such as in the methods to
  mark and unmark tasks

rahuljai-05 committed:
Merge branch 'branch-Level-5'

rahuljai-05 committed:
Add test case to ensure code functions as expected

TJW0911 commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,23 @@
+public class Event extends Deadline {
+    protected String deadlineTill;
```
Perhaps it would be a good idea to name this something like "endTime" since events generally have a start and end time instead of using deadlineTill.

TJW0911 commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,23 @@
+    @Override
+    public String toString() {
+        String taskStatus = "[E][" + (isDone ? "X" : " ") + "] ";
+        String from = " (from: " + deadline.replaceFirst("^from\\s+", "");
+        String to = "to: " + deadlineTill.replaceFirst("^to\\s+", "") + ")";
+        return taskStatus + taskName + from + to;
```
Using deadline and deadlineTill to represent the start and end time seems a bit confusing for the reader. Maybe it would be a good idea to have Event extend Task instead.

TJW0911 commented on src/main/java/Fenix.java:
```diff
@@ -0,0 +1,197 @@
+        }
+    }
+
+    private void unmarkAsDone(String[] words) {
```
unmarkAsDone sounds a bit confusing, perhaps you could use markTaskAsUndone instead.

TJW0911 commented on src/main/java/Fenix.java:
```diff
@@ -0,0 +1,197 @@
+        System.out.println(HORIZONTAL_LINE_FENIX_MODIFICATION);
+    }
+
+    public void unmarkTaskAsDone(int taskNumber) {
```
Similarly unmarkTaskAsDone sounds a bit confusing, perhaps you could use markTaskAsUndone instead.

rahuljai-05 committed:
Change the data structure for storing tasks

Implemented the usage of the Java Collections classes.
Replaced the primitive normal array which was used for
storing tasks with the ArrayList<Task> collection which
makes it easier to find, add, delete tasks and also
dynamically change the size of the array

rahuljai-05 committed:
Modify user input processing loop and delete dead code

Deleted two methods and replaced their functionalities with
new code under the processUserInput method

rahuljai-05 committed:
Delete tasks

Added functionality to delete tasks,
extracted a getTaskIndex method
and deleted redundant variable taskNumber.
Also modified the test case in accordance with the string updates made

rahuljai-05 committed:
Final code cleanup for Level-6

rahuljai-05 committed:
Implement the basic operations for writing to a file

Added class FileHandler to take control of the file writing operations

rahuljai-05 committed:
Save tasks

- Added function to load data (retrieve and decipher info to store in task array)
  from the text file whenever the chatbot starts up
- Added function to save all the modified changes to the text file whenever the
  user says bye and closes the program
- Rewrote the toString() method for all the task types and added an extra parameter to
  the task class to allow the program to print the task descirptions
  differently when the task is being read from the saved text file and
  when it has just been created during that session

rahuljai-05 committed:
Merge branch 'branch-Level-6'

rahuljai-05 committed:
Merge branch 'branch-Level-7'

Resolved merge conflicts

rahuljai-05 committed:
Implement changes to ensure jar file functionality

The code does not save the tasks properly when running the jar file
generated from the previous commit. Therefore, added a new
method under the FileHandler class to ensure that the parent directory
and the text file for saving the task data exists before writing any data
to it. In the case where the file or the directory doesn't exist, they are
created accordingly.

rahuljai-05 committed:
Create more OOP classes

rahuljai-05 committed:
Separate the different class concerns

Divided the funtionalities between the different classes more strictly.
Each class such as the Parser, UI, TaskHandler and Storage can now
only access the main Fenix controller class and cannot interact with
each other as freely anymore. This helps improve modularity and in
the separation of concerns as the different classes are not as tightly
coupled anymore.

rahuljai-05 committed:
Implement find function

rahuljai-05 committed:
Import List to make make the showAllTasks function work

rahuljai-05 committed:
Print line after displaying the tasks output of find function

rahuljai-05 committed:
Define new constant in SampleStrings to make the previous commit work

rahuljai-05 committed:
Make a function to display the number of tasks yet to be done

rahuljai-05 committed:
Show number of tasks yet to be done after each Fenix modification and refactor

rahuljai-05 committed:
Add JavaDoc comments for Fenix class

rahuljai-05 committed:
Add JavaDoc comments for Parser class

rahuljai-05 committed:
Add JavaDoc comments for TaskHandler class

rahuljai-05 committed:
Add JavaDoc comments for UserInterface class

rahuljai-05 committed:
Add JavaDoc comments for FileHandler and Storage classes

rahuljai-05 committed:
Add JavaDoc comments for the Task class and its subclasses

rahuljai-05 committed:
Add JavaDoc comment for Main class

rahuljai-05 committed:
Merge pull request #1 from rahuljai-05/branch-Level-9

branch-Level-9

rahuljai-05 committed:
Resolve merge conflicts

rahuljai-05 committed:
Merge pull request #2 from rahuljai-05/branch-A-JavaDoc

Branch-A-JavaDoc

rahuljai-05 committed:
Add User Guide
